### PR TITLE
1955 om supply cold chain temperature sensors

### PR DESCRIPTION
**Pull request by andreievg**
Links to #1955 

# 👩🏻‍💻 What does this PR do? 

I'll try to add here as i review it

# 🧪 How has/should this change been tested? 

I'll try to add here as i review it

## 💌 Any notes for the reviewer?

I tried to break this PR down based on areas, but realised it would be very hard if I want to preserve commits, so decided to just PR the whole lot.

@adrianwb, usually if I was implementing something that touches the domains I would break it down like this (I tried suggesting it with the task breakdown in the issue):

* Database schema and diesel (would keep the query repository very light at this stage, just the skeleton)
* Sync stuff (change log, translations, sync unit test)
* Front end (how does it look, what API does it need, do mock API)
* Service (business logic)
* GraphQL (mapping for business logic)
* Sync Integration tests

You'll notice that front end is done before api or service, this is because of personal preference of keeping presentation layer as light as possible, and provide it with the API that's easy to consume (also it's a lot easier to do test in back end thus the more logic there the better)

I've added new features branch for this work

### CODE REVIEW

File: server/service/src/sync/translations/sensor.rs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    #[serde(rename = "logInterval")]
+    pub log_interval: Option<i32>,
+    pub is_active: bool,
+    #[serde(rename = "lastConnectionDate")]
+    #[serde(deserialize_with = "zero_date_as_option")]
+    #[serde(serialize_with = "date_option_to_isostring")]
+    pub last_connection_date: Option<NaiveDate>,
+    #[serde(rename = "lastConnectionTime")]
+    #[serde(deserialize_with = "naive_time")]
+    pub last_connection_time: NaiveTime,
```
Comment (andreievg) [inline server/service/src/sync/translations/sensor.rs:54]: Hmm, i have to remember why date and time differ (one is optional), will try to get back to you on that one

File: server/graphql/core/src/loader/sensor.rs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+use async_graphql::dataloader::*;
+use async_graphql::*;
+use std::collections::HashMap;
+
+pub struct SensorByIdLoader {
+    pub connection_manager: StorageConnectionManager,
+}
+
+#[async_trait::async_trait]
+impl Loader<String> for SensorByIdLoader {
```
Comment (andreievg) [inline server/graphql/core/src/loader/sensor.rs:16]: Btw, would need this only when it's a field in a collection, for example for locations

File: server/graphql/core/src/simple_generic_errors.rs [hunk 3]
```diff
@@ -267,6 +267,7 @@ impl InvoiceLineBelongsToAnotherInvoice {
 #[graphql(rename_items = "camelCase")]
 pub enum UniqueValueKey {
     Code,
+    Serial,
```
Comment (andreievg) [inline server/graphql/core/src/simple_generic_errors.rs:270]: Nice

File: server/graphql/sensor/src/mutations/insert.rs [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    DatabaseError(DatabaseError),
+}
+
+fn map_error(error: ServiceError) -> Result<InsertSensorErrorInterface> {
+    use StandardGraphqlError::*;
+    let formatted_error = format!("{:#?}", error);
+
+    let graphql_error = match error {
+        // Standard Graphql Errors
+        ServiceError::SensorAlreadyExists => BadUserInput(formatted_error),
```
Comment (andreievg) [inline server/graphql/sensor/src/mutations/insert.rs:95]: Just a quick brief about error handling in graphql (this should definitly be in our onboarding docs). We have standard and structured errors, standards error are the one that are return by Result and structured error are built into [schema via unions](https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces/). 

Front end doesn't know what type of standard errors exist, and thus they are meant to be there for edge cases that are outside of apps controll and that can't be actioned on by the user or don't need to be translated, an example is internal database error, or bad user input (which is when api is used in a wrong way due to bug etc..)

Structured errors are known to front end through introspection (graphql schemas are turned to typescript with codegen), and thus we can quite easily handle the different variations of errors, translate them etc.. and if api is updated, this is straight away caught during compilation (we can even do [narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html) on error types ). So a structured error woudl be something that needs translating or something we can action on:

https://github.com/openmsupply/open-msupply/blob/6131f65381868ccc0ad4a125940b4b3e39491596/client/packages/system/src/Stock/Components/Repack/RepackModal.tsx#L119-L125

https://github.com/openmsupply/open-msupply/blob/6131f65381868ccc0ad4a125940b4b3e39491596/server/graphql/repack/mutations/insert.rs#L102-L106

**Review by andreievg — COMMENTED** [review]
Thanks @adrianwb, super imperssed by how you managed to figure all of this out ! 

If you update from develop you'll get huge test speed improvements that have recently been done by @clemens-msupply 

Something else you can have a look at is [integration sync tests](https://github.com/openmsupply/open-msupply/blob/6131f65381868ccc0ad4a125940b4b3e39491596/server/service/src/sync/test/integration/README.MD)

I have a few general comments and one request.

In general, I think it's a good idea to only implement things that are being used, this help with review and reduces cognitive load for future developlment. Athough we do have quite a few TODO's in comment in the repo, we do try to capture them in issues (can just add on to the epic or to carry over issues). I've noticed you had commented out bits of code (like stock field in sensor end point, btw for that one I woudl assume relation from sensor to location and then to stock, unless sensor stock is the stock that was affected by temperature reading from the sensor, which would be a lot harder because of location movements for stock and for sensor).

Also noticed a insert/update/delete is being exposed in API, i think this would have been a good learning experience, but I don't see it being needed (or at least that work is not in the pipeline for a long time). I think sensors should be generated by back end (when mSupply cold chain app is synced, or when back end reads fridge tag file). I think there would only be one mutation for now, and that's to update location for sensor (maybe update location, name and is_active). 

The way I see it, there is API for front end to upload a fridget tag temperature file (for now can just be manually selected), and once uploaded back end reads it and processes it in service layer and upserts sensor (it woudl probably be rest API for this, do you want someone to have a go at adding it, or would you be comfortable to do it yourself ?), that rest end point can return an id for sensor (or logs, or log date range) that can then be queried through GraphqlAPI. The next step would be for client to auto find this file and maybe auto upload it, can even do notification after auto upload ¯\_(ツ)_/¯ (but I think manual upload would be sufficient for awhile).

Anyways I didn't actually review the mutation bits, I think we need to take them out (and that's the only request I have for now)

Would be great if you can put futuer commits in another branch (that's branched from this one, to make sure this PR doesn't get bigger)

Thank you

**Comment by adrianwb** [comment]
> Thanks @adrianwb, super imperssed by how you managed to figure all of this out !

Thanks for all this - I very much appreciate your help and guidance. You know what they say about standing on the shoulders of giants :-)

> If you update from develop you'll get huge test speed improvements that have recently been done by @clemens-msupply

:-)

> Something else you can have a look at is [integration sync tests](https://github.com/openmsupply/open-msupply/blob/6131f65381868ccc0ad4a125940b4b3e39491596/server/service/src/sync/test/integration/README.MD)

Noted
 
> I have a few general comments and one request.
> 
> In general, I think it's a good idea to only implement things that are being used, this help with review and reduces cognitive load for future developlment. Athough we do have quite a few TODO's in comment in the repo, we do try to capture them in issues (can just add on to the epic or to carry over issues). I've noticed you had commented out bits of code (like stock field in sensor end point, btw for that one I woudl assume relation from sensor to location and then to stock, unless sensor stock is the stock that was affected by temperature reading from the sensor, which would be a lot harder because of location movements for stock and for sensor).

Yeah, there's probably some unnecessary fluff in there - the idea for some of it at least was just so that I'd know where to add the breach & temperature log checks into sensors, but I'll go through that and do a cull

> Also noticed a insert/update/delete is being exposed in API, i think this would have been a good learning experience, but I don't see it being needed (or at least that work is not in the pipeline for a long time). I think sensors should be generated by back end (when mSupply cold chain app is synced, or when back end reads fridge tag file). I think there would only be one mutation for now, and that's to update location for sensor (maybe update location, name and is_active).

Yes, almost all should only be editable on the back-end - on the front end it would just be location and is_active, and possibly name and store?

> The way I see it, there is API for front end to upload a fridget tag temperature file (for now can just be manually selected), and once uploaded back end reads it and processes it in service layer and upserts sensor (it woudl probably be rest API for this, do you want someone to have a go at adding it, or would you be comfortable to do it yourself ?), that rest end point can return an id for sensor (or logs, or log date range) that can then be queried through GraphqlAPI. The next step would be for client to auto find this file and maybe auto upload it, can even do notification after auto upload ¯_(ツ)_/¯ (but I think manual upload would be sufficient for awhile).

I don't mind having a go at this, but it'll take longer for me than for someone who has half of an idea of what they're doing! Bear in mind that I'm on holiday the first 2 weeks of September, so that may be a factor if there is anyone desperately waiting for this @adamdewey ?

> Anyways I didn't actually review the mutation bits, I think we need to take them out (and that's the only request I have for now)
> 
> Would be great if you can put futuer commits in another branch (that's branched from this one, to make sure this PR doesn't get bigger)

Will do

File: server/service/src/sync/translations/sensor.rs (hunk 1 above)
Comment (adrianwb) [reply server/service/src/sync/translations/sensor.rs:54]: Yeah, wasn't sure about that either - just did it the same way as `[transact]confirm_date/time`

File: server/repository/src/db_diesel/sensor_row.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        Ok(result)
+    }
+
+    pub fn find_many_by_id(&self, ids: &[String]) -> Result<Vec<SensorRow>, RepositoryError> {
+        Ok(sensor_dsl::sensor
+            .filter(sensor_dsl::id.eq_any(ids))
+            .load(&self.connection.connection)?)
+    }
+
+    pub fn delete(&self, id: &str) -> Result<(), RepositoryError> {
```
[the remark this answers is not in the document]
Comment (adrianwb) [reply server/repository/src/db_diesel/sensor_row.rs:110]: Yes, of course - definitely overkill

File: server/repository/src/migrations/v1_02_00/sensor.rs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        sql!(
+            connection,
+            r#"
+                ALTER TYPE changelog_table_name ADD VALUE IF NOT EXISTS 'sensor';
+            "#
+        )?;
+
+        sql!(
+            connection,
```
[the remark this answers is not in the document]
Comment (adrianwb) [reply server/repository/src/migrations/v1_02_00/sensor.rs]: Done in https://github.com/openmsupply/open-msupply/pull/2098/commits/40ab599abf3e38f6573dfaac8500a483b6b0dbe2 - just replicated your suggested changes in the main branch (I presume that "location" should be "sensor" :-) )

File: server/repository/src/db_diesel/sensor_row.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrianwb) [reply server/repository/src/db_diesel/sensor_row.rs]: Removed here: https://github.com/openmsupply/open-msupply/pull/2098/commits/1e500ddaf777abaaaf817ac76f5e32da2f3b2e36

File: server/repository/src/migrations/v1_02_00/sensor.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreievg) [reply server/repository/src/migrations/v1_02_00/sensor.rs]: >  just replicated your suggested changes in the main branch
🙏 

(presume that "location" should be "sensor" :-) )

🤦

File: server/service/src/sync/translations/sensor.rs (hunk 1 above)
Comment (andreievg) [reply server/service/src/sync/translations/sensor.rs:54]: > will try to get back to you on that one

Added to my list of TODO, won't quite look at it now

File: server/repository/src/db_diesel/user_permission_row.rs [hunk 7]
```diff
@@ -29,6 +29,8 @@ pub enum Permission {
     StoreAccess,
     // location,
     LocationMutate,
+    // sensor,
+    SensorMutate,
```
[the remark this answers is not in the document]
Comment (adrianwb) [reply server/repository/src/db_diesel/user_permission_row.rs:33]: Ah, that is helpful - I've added in SensorQuery (bound to ViewSensorDetails permission) and bound SensorMutate to EditSensorLocation permission here: https://github.com/openmsupply/open-msupply/pull/2098/commits/08760fd19b49ec2c80a17a9c2405034e8dd499e0

I'll revisit SensorMutate and see if I can restrict updating to just those fields we want to allow. FYI in classic mSupply, there's a bit of code which runs when you set the location of a sensor, to update the location for any breaches/temperature logs associated with it -> probably need to replicate that here somewhere...

File: server/service/src/sync/translations/sensor.rs [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    fn try_translate_pull_upsert(
+        &self,
+        _: &StorageConnection,
+        sync_record: &SyncBufferRow,
+    ) -> Result<Option<IntegrationRecords>, anyhow::Error> {
+        if !match_pull_table(sync_record) {
+            return Ok(None);
+        }
+
+        let data = serde_json::from_str::<LegacySensorRow>(&sync_record.data)?;
```
[the remark this answers is not in the document]
Comment (adrianwb) [reply server/service/src/sync/translations/sensor.rs]: I re-ran cargo fmt to fix the linting here: https://github.com/openmsupply/open-msupply/pull/2098/commits/e499cff08fabe391ab5fbaa02ed6be4e77587d64
And applied your simplified code suggestion here: https://github.com/openmsupply/open-msupply/pull/2098/commits/8f3a04f8927b8396dde104d689f820203e32bfb9

**Review by andreievg — COMMENTED** [review]
Thanks for the changes @adrianwb, looks great so far

> Thanks for all this - I very much appreciate your help and guidance. You know what they say about standing on the shoulders of giants :-)

Wow, thanks ! I would assume it's another joke though, something about Andrei the giant.. the french wrestler..

> Yeah, there's probably some unnecessary fluff in there - the idea for some of it at least was just so that I'd know where to add the breach & temperature log checks into sensors, but I'll go through that and do a cull

No probs, I just wanted to mention to make sure we are on the same page (in terms of the bits that edit sensors)

> Yes, almost all should only be editable on the back-end - on the front end it would just be location and is_active, and possibly name and store?

Yep, I added a new task in the epic (to clean up any unused mutations, we can do this at the end, woudl be easier to see what we actually need then)

> I don't mind having a go at this, but it'll take longer for me than for someone who has half of an idea of what they're doing! Bear in mind that I'm on holiday the first 2 weeks of September, so that may be a factor if there is anyone desperately waiting for this @adamdewey ?

How about we see how far you get by then ? And maybe while you away, we can do the tricky bits ?

**Comment by andreievg** [comment]
I approve this PR btw, but get this (so I guess you can approve and merged into the base feature branch !

Screenshot 2023-08-21 at 6 35 30 PM

**Comment by adrianwb** [comment]
> Thanks for the changes @adrianwb, looks great so far
> 
> > Thanks for all this - I very much appreciate your help and guidance. You know what they say about standing on the shoulders of giants :-)
> 
> Wow, thanks ! I would assume it's another joke though, something about Andrei the giant.. the french wrestler..

https://www.science.edu/acellus/2018/11/newton-seeing-further/

> > Yeah, there's probably some unnecessary fluff in there - the idea for some of it at least was just so that I'd know where to add the breach & temperature log checks into sensors, but I'll go through that and do a cull
> 
> No probs, I just wanted to mention to make sure we are on the same page (in terms of the bits that edit sensors)
> 
> > Yes, almost all should only be editable on the back-end - on the front end it would just be location and is_active, and possibly name and store?
> 
> Yep, I added a new task in the epic (to clean up any unused mutations, we can do this at the end, woudl be easier to see what we actually need then)

Sounds good

> > I don't mind having a go at this, but it'll take longer for me than for someone who has half of an idea of what they're doing! Bear in mind that I'm on holiday the first 2 weeks of September, so that may be a factor if there is anyone desperately waiting for this @adamdewey ?
> 
> How about we see how far you get by then ? And maybe while you away, we can do the tricky bits ?

Will do :-) I'll try to make as much progress as I can on adding in breaches & logs to the back end, and leave the UI side for the meantime

**Review by adrianwb — APPROVED** [review]
Approving as a proxy for Andrei who did the  real review!

**Pull request merged**